manager/controllers/app/blueprint_controller.go

```diff
@@ -51,41 +52,87 @@ func (r *BlueprintReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	if !blueprint.DeletionTimestamp.IsZero() {
+		// The object is being deleted
 		log.V(0).Info("Reconcile: Deleting Blueprint " + blueprint.GetName())
-	} else {
-		log.V(0).Info("Reconcile: Installing/Updating Blueprint " + blueprint.GetName())
+		return ctrl.Result{}, nil
 	}
 
+	observedStatus := blueprint.Status.DeepCopy()
+	log.V(0).Info("Reconcile: Installing/Updating Blueprint " + blueprint.GetName())
+
 	result, err := r.reconcile(ctx, log, &blueprint)
 	if err != nil {
 		return ctrl.Result{}, errors.Wrap(err, "failed to reconcile blueprint")
 	}
 
-	if blueprint.DeletionTimestamp == nil {
-		if !equality.Semantic.DeepEqual(&blueprint.Status, observedStatus) {
-			if err := r.Client.Status().Update(ctx, &blueprint); err != nil {
-				return ctrl.Result{}, errors.WrapWithDetails(err, "failed to update blueprint status", "status", blueprint.Status)
-			}
+	if !equality.Semantic.DeepEqual(&blueprint.Status, observedStatus) {
+		if err := r.Client.Status().Update(ctx, &blueprint); err != nil {
+			return ctrl.Result{}, errors.WrapWithDetails(err, "failed to update blueprint status", "status", blueprint.Status)
 		}
 	}
 
 	log.Info("blueprint reconcile cycle completed", "result", result)
 	return result, nil
 }
 
-func (r *BlueprintReconciler) deleteChartResource(ctx context.Context, kubeNamespace string, releaseName string) (ctrl.Result, error) {
-	rel, err := r.Helmer.Status(kubeNamespace, releaseName)
-	if err == nil && rel != nil {
-		_, _ = r.Helmer.Uninstall(kubeNamespace, releaseName)
+// reconcileFinalizers reconciles finalizers for Blueprint
+func (r *BlueprintReconciler) reconcileFinalizers(blueprint *app.Blueprint) error {
+	// finalizer
+	finalizerName := r.Name + ".finalizer"
+	hasFinalizer := utils.ContainsFinalizer(blueprint.GetFinalizers(), finalizerName)
+
+	// If the object has a scheduled deletion time, delete it and its associated resources
+	if !blueprint.DeletionTimestamp.IsZero() {
+		// The object is being deleted
+		if hasFinalizer { // Finalizer was created when the object was created
+			// the finalizer is present - delete the allocated resources
+			if err := r.deleteExternalResources(blueprint); err != nil {
+				return err
+			}
+
+			// remove the finalizer from the list and update it, because it needs to be deleted together with the object
+			ctrlutil.RemoveFinalizer(blueprint, finalizerName)
+
+			if err := r.Update(context.Background(), blueprint); err != nil {
+				return err
+			}
+		}
+		return nil
 	}
+	// Make sure this CRD instance has a finalizer
+	if !hasFinalizer {
+		ctrlutil.AddFinalizer(blueprint, finalizerName)
+		if err := r.Update(context.Background(), blueprint); err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
-	return ctrl.Result{}, nil
+func getReleaseName(step app.FlowStep) string {
+	// we add the "r" character at the beginning of the release name, since it must begin with an alphabetic character
+	return "r" + utils.Hash(step.Name, 20)
```

Comment:
unrelated to this PR but I think that we should start moving common code like `Hash` to appropriate folders in `pkg`